src/github.com/getlantern/connpool/connpool_test.go

```diff
@@ -94,38 +95,69 @@ func TestDialFailure(t *testing.T) {
 	if err != nil {
 		t.Fatalf("Unable to start test server: %s", err)
 	}
-	p := &Pool{
-		MinSize:              10,
-		RedialDelayIncrement: 10 * time.Millisecond,
-		MaxRedialDelay:       100 * time.Millisecond,
+
+	_, fdc, err := fdcount.Matching("TCP")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	poolSize := 10
+
+	p := New(Config{
+		Size: poolSize,
 		Dial: func() (net.Conn, error) {
 			atomic.AddInt32(&dialAttempts, 1)
 			if fail == int32(1) {
-				return nil, fmt.Errorf("I'm failing!")
+				return nil, fmt.Errorf("I'm failing intentionally!")
 			}
 			return net.DialTimeout("tcp", addr, 15*time.Millisecond)
 		},
-	}
+	})
 
-	p.Start()
-	defer p.Stop()
+	// Try to get connection, make sure it fails
+	conn, err := p.Get()
+	if !assert.Error(t, err, "Dialing should have failed") {
+		conn.Close()
+	}
 
 	// Wait for fill to run for a while with a failing connection
 	time.Sleep(1 * time.Second)
-	assert.True(t, dialAttempts < 500, fmt.Sprintf("Should have had a small number of dial attempts, but had %d", dialAttempts))
+	assert.Equal(t, 1, atomic.LoadInt32(&dialAttempts), fmt.Sprintf("There should have been only 1 dial attempt"))
+	assert.NoError(t, fdc.AssertDelta(0), "There should be no additional file descriptors open")
 
 	// Now make connection succeed and verify that it works
 	atomic.StoreInt32(&fail, 0)
 	time.Sleep(100 * time.Millisecond)
 	connectAndRead(t, p, 1)
 
+	time.Sleep(fillTime)
+	log.Debug("Testing")
+	assert.NoError(t, fdc.AssertDelta(10), "Pool should have filled")
+
 	// Now make the connection fail again so that when we stop, we're stopping
 	// while failing (tests a different code path for stopping)
 	atomic.StoreInt32(&fail, 1)
 	time.Sleep(100 * time.Millisecond)
+
+	p.Close()
+
+	assert.NoError(t, fdc.AssertDelta(0), "All connections should be closed")
+}
+
+func TestPropertyChange(t *testing.T) {
+	claimTimeout := 5 * time.Second
+
+	cfg := &Config{
+		ClaimTimeout: claimTimeout,
+	}
+	p := New(*cfg)
+	defer p.Close()
+
+	cfg.ClaimTimeout = 7 * time.Second
+	assert.NotEqual(t, claimTimeout, cfg.ClaimTimeout, "Property changed on config shouldn't be reflected in pool")
```

Comment:
I fail to see how this assertion proves that `p`'s configuration was in fact not affected, as we are only comparing `claimTimeout` and `cfg`. `p` must be not affected as we're using a copy of Config, but if we wanted to test it I think that casting `p` to `*pool` and checking if the initial value of `ClaimTimeout` is still equal to `claimTimeout` would give better results:

```go
assert.Equal(t, claimTimeout, p.(*pool).ClaimTimeout, "Property changed on config shouldn't be reflected in pool")
```

or using `cfg.ClaimTimeout` instead of `claimTimeout`:

```go
assert.NotEqual(t, cfg.ClaimTimeout, p.(*pool).ClaimTimeout, "Property changed on config shouldn't be reflected in pool")
```